### DISCUSSION TITLE
Arbitrary filters through kitten?

### DISCUSSION THREAD

**Discussion by salmankhilji** [kovidgoyal/kitty#5912] [Ideas]
What if we allow `kitten` to run an arbitrary filter on the host machine? i.e., Run the given program on the host computer with these arguments and output the result to stdout.

### Use Case #1

While running on embedded systems, such as Android, basic tools like `awk` and `sed` are often provided by `busybox` or `toybox`.  These replacements are often limited and buggy.  If we could run an arbitrary filter through `kitten`, we could do:

```sh
alias awk='kitten filter awk'
alias sed='kitten filter sed'
```

Now, we can run our local `awk` and `sed` on remote machines with all the bells and whistles.

### Use Case #2

When running `fzf` locally, I like to specify a _darker_ or _lighter_ shade of the terminal background color.  This is what I currently do:

```
color="$(kitty @ get-colors | sed -nr 's/^background\s+(.+)/\1/p' | xargs pastel darken 0.05 | pastel format)"
```

While this works locally on my machine, I would like to retain this capability within a script running on my Android device (or even SSH).

Luckily, `pastel` happens to provide pre-builts for `arm` so I could this on Android, but what if there wasn't a pre-built binary?  Or what if I didn't want to have to install `pastel` on a remote host?  The proposed feature in `kitten` would rid the need to have to manually cross compile such tools for embedded systems.

Another program that comes to mind is `column`.  While this basic tool is available on most Linux systems, I have yet to find a pre-built for Android/arm.

### Security Risks

Running an arbitrary program on our local machines from remote hosts is obviously a security risk.  By running the following on the remote host:

```
kitten filter rm -rf ~/Documents
```

one could easily blow way important data on my local machine, so we may need to allow only a white-listed subset of filters that are specified in, say, `kitty.conf`.  (GNU `sed` allows arbitrary program execution through the `-e` flag, so adding `sed` to the white-list would still be a security risk.)

### Background

Initially, I thought of requesting a `pastel` like kitten that simply returned a subset of what `pastel` provides—primarily the `lighten` and `darken` commands.  Then I thought that perhaps a more generic filter may allow other use cases.

**Comment by kovidgoyal** [comment]
I'm not a fan of this concept. It's got a huge security footprint and a
is only useful on underpowered remote systems, which is a niche use
case.

For your specific usecase you can simply write a script that invokes the
ssh kitten while setting a env var with the lighter and darker colors
calculated on the local machine. Your remote script can use those env
vars. Or you could even just have those env vars set in bashrc and have
the ssh kitten forward them to the remote computer automatically.

**Reply by salmankhilji** [reply]
> I'm not a fan of this concept. It's got a huge security footprint and a is only useful on underpowered remote systems, which is a niche use case. For your specific usecase you can simply write a script that invokes the ssh kitten while setting a env var with the lighter and darker colors calculated on the local machine. Your remote script can use those env vars. Or you could even just have those env vars set in bashrc and have the ssh kitten forward them to the remote computer automatically.

Yes, this is what I currently do:

```sh
setup_shell() {
  local background
  background="$(kitty @ get-colors | sed -nr 's/^background\s+(.+)/\1/p' | xargs pastel darken 0.05 | pastel format)"
  adb shell "
  print 'stty cols $(tput cols)\nstty rows $(tput lines)' >/data/.resize
  print 'COLOR_FZF=$background' >/data/.fzf-theme
  grep -qE '/data/\.mkshrc' /system/etc/mkshrc || print '\n[ -f /data/.mkshrc ] && . /data/.mkshrc' >>/system/etc/mkshrc
  "
}
```

The `adb shell` command runs the provided script on the Android device, which BTW has `mksh`, not `bash`.
